### PR TITLE
Fix #318: Set up code coverage

### LINKED ISSUE #318: Set up code coverage
It would be nice to see how good our tests are.

I've used [scoverage](https://github.com/scoverage) a bit so maybe it's worth it to start there.

It would be *really* fancy to have Travis automatically calculate code coverage for us. One site I've seen that does this is [https://codecov.io/](https://codecov.io/). The Miniboxing project is using this right now and it seems interesting.

### PR DESCRIPTION
**Pull request by InTheNow**
@non This will require that you add non/cats to https://codecov.io/

@travisbrown - does this look OK?

There are issues with scalajs and scoverage, hence new PR to master

### CODE REVIEW

**Comment by non** [comment]
Looks good to me. :+1: 

I assume `bash <(curl -s https://codecov.io/bash)` is the preferred way to hook this up with Travis? A bit freaky, but if Travis is OK with it then I am as well.

**Comment by InTheNow** [comment]
> I assume 

see https://codecov.io/guide#scala

File: build.sbt [hunk 1]
```diff
@@ -7,6 +7,17 @@ import sbtrelease.ReleasePlugin.ReleaseKeys.releaseProcess
 import sbtrelease.ReleaseStateTransformations._
 import sbtrelease.Utilities._
 import sbtunidoc.Plugin.UnidocKeys._
+import ScoverageSbtPlugin._
+
+lazy val scoverageSettings = Seq(
+  ScoverageKeys.coverageMinimum := 60,
+  ScoverageKeys.coverageFailOnMinimum := false,
+  ScoverageKeys.coverageHighlighting := {
+    if (scalaBinaryVersion.value == "2.10") false
```
Comment (travisbrown) [inline build.sbt]: This is super nitpicky, but I've always thought of `if (cond) false else true`-ish stuff as a code smell. Why not just `scalaBinaryVersion.value != "2.10"`?

File: .travis.yml [hunk 2]
```diff
@@ -2,7 +2,7 @@ language: scala
 scala:
   - 2.11.6
 script:
-  - sbt ++$TRAVIS_SCALA_VERSION validate publishLocal
+  - sbt ++${TRAVIS_SCALA_VERSION} coverage validate coverageReport coverageAggregate && bash <(curl -s https://codecov.io/bash) && sbt clean validate publishLocal ;
```
Comment (travisbrown) [inline .travis.yml]: Just curious—why add brackets here?

File: build.sbt [hunk 3]
```diff
@@ -7,6 +7,17 @@ import sbtrelease.ReleasePlugin.ReleaseKeys.releaseProcess
...
+    else true
+  },
+  ScoverageSbtPlugin.ScoverageKeys.coverageExcludedPackages := "cats\\.bench\\..*"
```
Comment (travisbrown) [inline build.sbt]: Might as well take advantage of the fact that you've imported the contents of `ScoverageSbtPlugin`.

File: .travis.yml (hunk 2 above)
Comment (travisbrown) [inline .travis.yml]: Has sbt-scoverage found a way to support `coverageAggregate` in the same SBT invocation? The [README](https://github.com/scoverage/sbt-scoverage#multi-project-reports) still says it needs to be in a separate one.

Comment (InTheNow) [inline .travis.yml]: > why add brackets here?

No particular reason, other than some bits that I did have and later removed, also used that style. I'll revert to no-bracket

> Has sbt-scoverage found a way to support coverageAggregate

I read that too but it seems to work fine - will have a closer inspection later

File: build.sbt (hunk 1 above)
Comment (InTheNow) [inline build.sbt]: Not picky at all - I'll change it

File: build.sbt (hunk 3 above)
Comment (InTheNow) [inline build.sbt]: Yes - I saw that right after I did the PR - why can one stare at something and only see things by pressing return? :)

**Comment by InTheNow** [comment]
One open issue (relates to the above mentioned code I deleted) is _when_ the stats should be updated? My thoughts are:

- Not for PR's
- Only for Snapshots

**Comment by InTheNow** [comment]
As for the _when_, always looks fine.

@non @travisbrown - From my side, all is done - see https://github.com/InTheNow/cats/tree/fix/318 for the working badge

**Comment by InTheNow** [comment]
@non @travisbrown - Any thoughts on when this might be merged, please? Only reason for asking is that I am finishing off the snapshot work and ideally I would like to rebase that PR on this as the same files are covered

**Comment by non** [comment]
Still :+1: from me. As soon as someone else signs-off I will merge.

**Comment by travisbrown** [comment]
:+1: from me as well.

**Pull request merged**